x-pack/auditbeat/module/system/host/host.go

```diff
@@ -5,23 +5,145 @@
 package host
 
 import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"io"
 	"net"
+	"strconv"
+	"time"
 
+	"github.com/OneOfOne/xxhash"
+	"github.com/joeshaw/multierror"
 	"github.com/pkg/errors"
 
+	"github.com/elastic/beats/auditbeat/datastore"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/common/cfgwarn"
-	"github.com/elastic/beats/metricbeat/mb"
-
 	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
 	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
 )
 
 const (
 	moduleName    = "system"
 	metricsetName = "host"
+
+	bucketName        = "host.v1"
+	bucketKeyLastHost = "lastHost"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+)
+
+type eventAction uint8
+
+const (
+	eventActionHost eventAction = iota + 1
+	eventActionIDChanged
+	eventActionReboot
+	eventActionHostnameChanged
+	eventActionHostChanged
 )
 
+func (action eventAction) string() string {
+	switch action {
+	case eventActionHost:
+		return "host"
+	case eventActionIDChanged:
+		return "host_id_changed"
+	case eventActionReboot:
+		return "reboot"
+	case eventActionHostnameChanged:
+		return "hostname_changed"
+	case eventActionHostChanged:
+		return "host_changed"
+	default:
+		return ""
+	}
+}
+
+// Host represents information about a host.
+type Host struct {
+	info types.HostInfo
+	// Uptime() in types.HostInfo recalculates the uptime every time it is called -
+	// so storing it permanently here.
+	uptime time.Duration
+	addrs  []net.Addr
+	macs   []net.HardwareAddr
+}
+
+// changeDetectionHash creates a hash of selected parts of the host information.
+// This is used later to detect changes to a host over time.
+func (host *Host) changeDetectionHash() uint64 {
+	h := xxhash.New64()
+
+	if host.info.Containerized != nil {
+		h.WriteString(strconv.FormatBool(*host.info.Containerized))
+	}
+
+	h.WriteString(host.info.Timezone)
+	binary.Write(h, binary.BigEndian, int32(host.info.TimezoneOffsetSec))
+	h.WriteString(host.info.Architecture)
+	h.WriteString(host.info.OS.Platform)
+	h.WriteString(host.info.OS.Name)
+	h.WriteString(host.info.OS.Family)
+	h.WriteString(host.info.OS.Version)
+	h.WriteString(host.info.KernelVersion)
+
+	return h.Sum64()
+}
+
+func (host *Host) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		// https://github.com/elastic/ecs#-host-fields
+		"uptime":              host.uptime,
+		"boottime":            host.info.BootTime,
+		"timezone.name":       host.info.Timezone,
+		"timezone.offset.sec": host.info.TimezoneOffsetSec,
+		"hostname":            host.info.Hostname,
+		"id":                  host.info.UniqueID,
+		// TODO "host.type": ?
```

Comment:
Can you put some more details about what needs to be done here with this TODO. And possibly link it back to some issue if there is some that tracks the remaining work. Or remove it.

Reply:
It referred to implementing the ECS field `host.type` - but I don't see what with so I'm removing it.